client/lib/wpcom-undocumented/lib/undocumented.js

```diff
@@ -787,9 +791,30 @@ Undocumented.prototype.cart = function( cartKey, method, data, fn ) {
 		method = 'GET';
 		data = {};
 	}
+
+	if ( method === 'POST' ) {
```

Comment:
I'm not a fan of overloading the method here but I see that it's already in use around the code but only in a couple places. what is your assessment in how much work it would take to replace calls like `cart( …, 'GET', … )` and `cart( …, 'POST', … )` with calls like `getCart( … )` and `setCart( … )`?

Reply:
Thanks for the review @dmsnell 👍 

Are you thinking that `getCart()` and `setCart()` would be added to `Undocumented.prototype`? If so, I see it being a very simple change as you noted and I'd be happy to implement that first and use it in this PR.

Would it be better if `getCart()` and `setCart()` were in [wpcom.js](https://github.com/Automattic/wpcom.js)? I'm pretty new to Calypso development so I'm learning from all of you here.

Reply:
If you think it's easy then go ahead and merge those changes into this PR. It could be worth a pre-emptive PR to do that and then rebase this work off of it. The expansion of the `POST` behavior is primarily what leads me to want to make the change.

They don't belong in `wpcom.js` - almost nothing does anymore - and we ideally want to make minimal changes here so I think `Undocumented` is still the best place to put this. Please also check with those who have worked on this cart code in the past - they may have much better context than I do :smile:

Reply:
Will do, thanks.

Reply:
@jaswrks It seems like a duplicate logic, there is already a code that updates cart:

https://github.com/Automattic/wp-calypso/blob/master/client/lib/cart/store/index.js
https://github.com/Automattic/wp-calypso/blob/master/client/lib/cart/store/cart-synchronizer.js

Every action that updates CartStore, e.g. `case UpgradesActionTypes.CART_COUPON_APPLY:` triggers CartSynchronizer to send a POST request to synchronize it with server side data. So it would be enough to issue some actions like `UpgradesActionTypes.CART_ITEMS_ADD` and `UpgradesActionTypes.CART_COUPON_APPLY` and POST requests are going to be sent.

Reply:
Thank you both!

I've rebased this on https://github.com/Automattic/wp-calypso/pull/25439 where `cart()` was divided into `getCart()` and `setCart()`. I also moved the coupon detection logic into `preprocessCartForServer` in `CartSynchronizer`.